### DISCUSSION TITLE
How to resize and compress in one go

### DISCUSSION THREAD

**Discussion by perason** [dlemstra/Magick.NET#613] [General] [answered]
Today I do this

    using (var stream = new System.IO.MemoryStream())
    {
        formfile.CopyTo(stream);

        using (ImageMagick.MagickImage image = new ImageMagick.MagickImage(stream))
        {
            if (image.Width > 1920 || image.Height > 1080)
            {
                if (image.Width > image.Height)
                {
                    image.Resize(1920, 0);
                }
                else
                {
                    image.Resize(0, 1080);
                }                                    
                image.Write(FileName);
                                   
                var optimizer = new ImageMagick.ImageOptimizer
                {
                    IgnoreUnsupportedFormats = true,
                };
                optimizer.Compress("changed_" + FileName));
            }
        }
    }

Is there any way to optimize this and do something like `image.Compress()` before saving it to disk.

**Edit**

As a note, I was thinking to first use the `stream` and compress it before resizing, though it felt kind of backwards doing it in that order. 

I mean, why to compress a bigger image than necessary, still, performance-wise, would it perform better and what would the end result be when it comes to size/quality?

**Edit 2**

I did a test, where I first compressed the image and then resized it, though it take both longer time and the size is bigger.

**Comment by dlemstra** [comment, marked as answer]
My advise would be to call `image.Write(yourOutputStream);` and then call `optimizer.Compress(yourOutputStream)` to compress the stream. The optimizer should always be the last step. And you could do `new ImageMagick.MagickImage(formfile` instead of reading it from a memory stream.

**Reply by perason** [reply]
Thanks. Will do that. 

But I can't pass an `IFormFile` to `new ImageMagick.MagickImage(formfile)` ...or did I miss something?

**Edit**

Do you mean to use `formfile.OpenReadStream()` instead of `formfile.CopyTo(stream);`, e.g.

    new ImageMagick.MagickImage(formfile.OpenReadStream())

**Reply by dlemstra** [reply]
I read it as `fromFile` instead of `formFile` and I assumed it was a file name. Maybe you could do something similar to this:

```C#
using (var input = formfile.OpenReadStream())
{
    using (var output = new MemoryStream())
    {
        using (var image = new MagickImage(input))
        {
            if (image.Width > 1920 || image.Height > 1080)
            {
                image.Resize(1920, 1080);
                image.Write(output);
            }
            else
            {
                input.Position = 0;
                await input.CopyToAsync(output);
            }
        }

        var optimizer = new ImageOptimizer
        {
            IgnoreUnsupportedFormats = true,
        };

        output.Position = 0;
        optimizer.Compress(output);

        using (var fileStream = File.OpenWrite(FileName))
        {
            await output.CopyToAsync(fileStream);
        }
    }
}
```

**Reply by perason** [reply]
Thanks again.

I still hope you consider to add the ability to use `MagickImage` and do e.g.

    using (var image = new MagickImage(input))
    {
        image.Resize(1920, 1080);
        image.Compress(new ImageOptimizerOptions { IgnoreUnsupportedFormats = true });
        image.Write(output);
    }

**Reply by dlemstra** [reply]
I will not add this because this means we need to encode and decode the image. A possible option could be `image.WriteCompressed(output);` but I will give this some thought.

**Reply by perason** [reply]
Great, that would work too.

I mean, if one need to do both, would such option improve/boost the overall performance, both on memory usage and speed, or else it works just fine as is.
